### ISSUE TITLE
Changing Keymap

### ISSUE THREAD

**Issue by lite20** [NodeOS/NodeOS#236]
NodeOS currently has no method of swapping keyboard layouts. This thread is to mark progress in adding this function to NodeOS. I suggest it be a NodeJS module as it could be useful to other programs, and not everyone will need it.

**Comment by Coretool** [comment]
I don't know if a npm module is the right choice for this task. I think the best way is giving it as kernel arg

**Comment by piranna** [comment]
Giving a kernel argument it's easy, maybe we could add a solution out of the box, for example a flag on the scripts that exec QEmu. Maybe could be a simple solution for SysLinux and the live CD images, too.

**Comment by Coretool** [comment]
We can pass kernel parameters via the `syslinux.cfg`... It should be possible to make changes via a node module (more of a command line app). If nobody else wants I can give it a try :) !

**Comment by piranna** [comment]
> We can pass kernel parameters via the syslinux.cfg...

Yes, but I was thinking more about to use a syslinux plugin, like the menu
one or something similar. We can left it for a second round and update
first the QEmu scripts.

>…It should be possible to make changes via a node module (more of a
command line app). If nobody else wants I can give it a try :) !

That will not work, but the QEmu scripts part is simple enough. They are
spread and duplicated on the several layers, so if you want to unify them
go for it :-)

We need to start tagging and managing the bug related issues...

**Comment by jaumard** [comment]
Generally I use `setxkbmap fr` to set my keyboard to AZERTY on linux, maybe an npm module can do the same to be use as command line ? Don't know if it's possible because I didn't understand yet how nodeOS works.

**Comment by Coretool** [comment]
`setxkbmap` is a `xlib` command/method so it won't work on nodeOs since we don't have X11 I think we will have to work with kernel args to achieve this..

**Comment by piranna** [comment]
> setxkbmap is a xlib command/method so it won't work on nodeOs since we don't have X11 I think we will have to work with kernel args to achieve this...

console-utils could done this, maybe...


> Can we set the keyboard layout while building nodeOs ? (When the syslinux.cfg gets created ?)

I once had it hardcoded with spanish keyboard, so it's possible :-) I removed it because it's better to add a flag, but we didn't do it yet.

**Comment by piranna** [comment]
If someone wants to do it please tell me, we need to start to organize collaborations and assign tasks to whom what to pick them so we don't duplicate efforts :-)

**Comment by Coretool** [comment]
I can do it ! How / where did you hardcode it ?

**Comment by piranna** [comment]
In the QEmu scripts on the `append` argument.

**Comment by Coretool** [comment]
That's what you would do in the syslinux.cfg too ...

**Comment by piranna** [comment]
Then there you are :-) Only that take in account `syslinux.cfg` will be on the CD image and will not be able to be change, so or you are an option to build it with a hardcoded keymap, or you add an option to select when booting the Live CD as Ubuntu install ones does...

**Comment by Coretool** [comment]
@piranna Which script does generate a the `syslinux.cfg` ?

**Comment by piranna** [comment]
> @piranna Which script generates the syslinux.cfg ?

[It's not generated](https://github.com/NodeOS/NodeOS/blob/master/node_modules/nodeos-rootfs/syslinux.cfg)

**Comment by Coretool** [comment]
Ah ! There it is ^^

**Comment by lite20** [comment]
> Giving a kernel argument it's easy, maybe we could add a solution out of the box, for example a flag on the scripts that exec QEmu. Maybe could be a simple solution for SysLinux and the live CD images, too.

Does NodeOS run on bare metal? It didn't work the last time I tried. What's the procedure for setting it up?

**Comment by lite20** [comment]
> Generally I use setxkbmap fr to set my keyboard to AZERTY on linux, maybe an npm module can do the same to be use as command line ? Don't know if it's possible because I didn't understand yet how nodeOS works.

Yeah, this is what I had in mind. Think of NodeOS as ubuntu server with v8 running everything. (Of course, it's more complex and streamlined than that, but that POV makes it fairly easy to help understand "would this work" questions)

**Comment by lite20** [comment]
> Generally I use setxkbmap fr to set my keyboard to AZERTY on linux, maybe an npm module can do the same to be use as command line ? Don't know if it's possible because I didn't understand yet how nodeOS works.

The only thing is that we would need the command, and all it's dependencies too which all boils back down to the kernel.

**Comment by piranna** [comment]
> Does NodeOS run on bare metal? It didn't work the last time I tried. What's the procedure for setting it up?

Of course it does :-) You can use one of the iso prebuild images, or build it with `PLATFORM=iso`. After that, you'll need to write it on a USB stick with `bin/installUsb`.


> The only thing is that we would need the command, and all it's dependencies too which all boils back down to the kernel.

I would try with the kernel boot argument, that we know it already works...

**Comment by Coretool** [comment]
I did some research, and as far as I know now, there easiest eay of doing it, is making some changes to `/etc/vconsole.cfg` (or something like this) to change the layout for the console...

**Comment by piranna** [comment]
I think it was by defining the `LANG=` parameter.

**Comment by Coretool** [comment]
Let me try that ...

**Comment by lite20** [comment]
> I think it was by defining the LANG= parameter.

But won't changing the language only occur on startup? Users wouldn't be able to change the keyboard layout from then on?

**Comment by lite20** [comment]
I thought the intended goal was to be able to swap layouts on the fly like on most OS's? That's why I suggested a module, so apps could change the keyboard layout if their functionality required it.

**Comment by Coretool** [comment]
@lite20 It's just one way of doing it. I don't know how it looks like with accessing /etc/ on nodeOS, the keyboard related configs are all found there...

**Comment by piranna** [comment]
> @lite20 It's just one way of doing it.

Exactly, it's the one I found to do it, a simple one.

> I don't know how it looks like with accessing /etc/ on nodeOS, the
keyboard related configs are all found there...

Each user has its own /etc folder, so it's a posibility to set
configurations there, but I don't know if they get applied in a per-user
basis... I think it would be feasable, but someone else would need to test
it. If it's just a matter of setting the LANG= environment variable or
others, logon and/or nodeos-mount-filesystems could be updated to set them
defined on /etc/logon.json, what do you think?

**Comment by Coretool** [comment]
I don't know if this is a debian/ubuntu specific feature, but I know that there is a `LANG` environment on both if them. Changing it, changes keyboard and lang for this session. If it's not platform specific, we can use it ...

**Comment by piranna** [comment]
As I said, setting it as boot argument worked for me, so seems to be
platform agnostic. It could be easy to fetch it from host system and set it
for QEmu scripts, but doing so for SysLinux and ISO images will be trickier
and it would need to show a menu to select the language on boot.

Any volunteer for any of the two things? :-)

**Comment by luii** [comment]
Maybe it would be a solution to hardcode the keymaps, for example map "a" to "b" or something
Using [KeyboardJS](https://github.com/RobertWHurst/KeyboardJS) for this or something similiar 

And creating a binary file for this [loadkeys](http://www.tutorialspoint.com/unix_commands/loadkeys.htm)

**Comment by piranna** [comment]
> Maybe it would be a solution to hardcode the keymaps, for example map "a" to "b" or something
> Using KeyboardJS for this or something similiar

That's too much high level, I'm not sure if it would be possible...


> And creating a binary file for this loadkeys

That would be interesting, yes.

**Comment by lite20** [comment]
@luii @piranna did we make any progress on this?

**Comment by luii** [comment]
not really, im too much into other things like docker support, `nodeos-mount-filesystems` and school so there was not much time for this :-/

**Comment by lite20** [comment]
@luii all good! I'm sure someone will pick it up, and if not, I can try when I finish my work stack.
